### PR TITLE
Implement XAI for tiling detection

### PR DESCRIPTION
**Pull request by GalyaZalesskaya**
### Summary

This PR implements the functionality to use saliency maps for tiles and merge them in the unite saliency map. The current approach is to resize a whole image and use its saliency map, which suffers from the lack of details. 
The results of the new approach and the current one are below.

image

Also, the exact same merging mechanism was applied for Mask RCNN (Instance Segmentation) Pytorch inference:
image

For Mask RCNN (Instance Segmentation) OpenVINO inference saliency maps aren't returned for each tole. So, the merged masks for the united image were used.
image

## What was added:
- Merging saliency maps functionality, using both saliency maps from tiles and a whole image, add non-linear normalization afterward
- Add implementation for PyTorch inference and OpenVINO inference for all detection models

- Add implementation to merge saliency maps for Pytorch task for Instance Segmentation task (Mask RCNN)
- Updated saliency map generation for OpenVINO inference task for Instance Segmentation
  - Changed a way to merge masks for each objects, added confidence scores to mask value
  - Moved saliency map generation process for Instance Segmentation from [/model_wrappers/openvino_models.py](https://github.com/openvinotoolkit/training_extensions/blob/a394a17c947f89fe5ac868949d907379b528da8a/otx/algorithms/detection/adapters/openvino/model_wrappers/openvino_models.py#L172) to [tiler.py](https://github.com/openvinotoolkit/training_extensions/blob/a394a17c947f89fe5ac868949d907379b528da8a/otx/api/utils/tiler.py#L4) implementation to keep tiled saliency map generation for OpenVINO task in one file (negotiable)

- Add parameter in BaseRecordingForwardHook whether to return normalized saliency maps 
- Add parameter in DetClassProbabilityMapHook whether to use softmax in saliency maps postprocessing

- Fixed the issue there during workplace creation for Mask RCNN model `deployment_tile_classifier.py` config file wasn't created

- Unit tests
- e2e tests

[CVS-108355](https://jira.devtools.intel.com/browse/CVS-108355)

### Checklist

- [x] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [x] I have added e2e tests for validation.
- [x] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [x] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [x] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

**Comment by eunwoosh** [comment]
Could you add and update unit test?

File: otx/algorithms/common/adapters/mmcv/hooks/recording_forward_hook.py [hunk 1]
```diff
@@ -39,13 +40,15 @@ class BaseRecordingForwardHook(ABC):
         module (torch.nn.Module): The PyTorch module to be registered in forward pass
         fpn_idx (int, optional): The layer index to be processed if the model is a FPN.
                                   Defaults to 0 which uses the largest feature map from FPN.
+        normalize (bool, optional): Whether to normalize the resulting saliency maps.
```
Comment (eunwoosh) [inline otx/algorithms/common/adapters/mmcv/hooks/recording_forward_hook.py]:
```suggestion
        normalize (bool): Whether to normalize the resulting saliency maps.
```
It shouldn't get `None` right?

File: otx/algorithms/common/adapters/mmcv/hooks/recording_forward_hook.py [hunk 2]
```diff
@@ -39,13 +40,15 @@ class BaseRecordingForwardHook(ABC):
...
     """
 
-    def __init__(self, module: torch.nn.Module, fpn_idx: int = -1) -> None:
+    def __init__(self, module: torch.nn.Module, fpn_idx: int = -1, normalize=True) -> None:
```
Comment (eunwoosh) [inline otx/algorithms/common/adapters/mmcv/hooks/recording_forward_hook.py]:
```suggestion
    def __init__(self, module: torch.nn.Module, fpn_idx: int = -1, normalize: bool =True) -> None:
```
How about to add type hint?

File: otx/algorithms/detection/adapters/mmdet/hooks/det_class_probability_map_hook.py [hunk 3]
```diff
@@ -29,15 +29,16 @@
 class DetClassProbabilityMapHook(BaseRecordingForwardHook):
     """Saliency map hook for object detection models."""
 
-    def __init__(self, module: torch.nn.Module) -> None:
-        super().__init__(module)
+    def __init__(self, module: torch.nn.Module, normalize=True, use_cls_softmax=True) -> None:
```
Comment (eunwoosh) [inline otx/algorithms/detection/adapters/mmdet/hooks/det_class_probability_map_hook.py]:
```suggestion
    def __init__(self, module: torch.nn.Module, normalize: bool =True, use_cls_softmax: bool =True) -> None:
```
How about to add type hint?

File: otx/algorithms/detection/adapters/mmdet/task.py [hunk 4]
```diff
@@ -499,6 +504,11 @@ def _export_model(
             if export_options["deploy_cfg"]["codebase_config"]["task"] != "Segmentation":
                 if "saliency_map" not in output_names:
                     output_names.append("saliency_map")
+            # disable softmax and normalization to merge saliency map for tiles and postprocess them altogether
+            tiling_detection = True if "tile_cfg" in cfg else False
```
Comment (eunwoosh) [inline otx/algorithms/detection/adapters/mmdet/task.py]:
```suggestion
            tiling_detection = "tile_cfg" in cfg
```

File: otx/api/utils/tiler.py [hunk 5]
```diff
@@ -123,26 +125,33 @@ def predict_sync(self, image: np.ndarray, tile_coords: List[List[int]]):
...
-                features = (
-                    copy.deepcopy(raw_predictions["feature_vector"].reshape(-1)),
-                    copy.deepcopy(raw_predictions["saliency_map"][0]),
+            # cache each tile feature vector and saliency map
+            if "feature_vector" in raw_predictions or "saliency_map" in raw_predictions:
+                tile_meta.update({"coord": coord})
+                features.append(
+                    (
+                        (
+                            copy.deepcopy(raw_predictions["feature_vector"].reshape(-1)),
```
Comment (eunwoosh) [inline otx/api/utils/tiler.py]: Is it necessary to `deepcopy` them?

**Review by eunwoosh — COMMENTED** [review]
Thanks for good enhancement! I left some comments. please take a look.

File: otx/algorithms/detection/adapters/mmdet/datasets/tiling.py [hunk 6]
```diff
@@ -502,3 +505,76 @@ def get_ann_info(self, idx):
...
+        Args:
+            feature_vectors: tile-level feature vectors.
+
+        Returns:
+            merged_vectors (list[np.ndarray]): Merged vectors for each image.
+        """
+
+        # tiles vectors + full image vector
+        vect_per_image = len(feature_vectors) // self.num_images + 1
+        image_vectors = feature_vectors[::vect_per_image]
```
Comment (jaegukhyun) [inline otx/algorithms/detection/adapters/mmdet/datasets/tiling.py]: Is this intended? If you don't want feature_vectors[:, :vect_per_image], then just resolve this conversation

**Comment by jaegukhyun** [comment]
Overally LGTM, As Eunwoo said, could you add unit test to cover your PR? And is there no need to add intg or e2e test for this feature?

**Comment by negvet** [comment]
Thank you for a PR!
Could you please rebase on develop (https://github.com/openvinotoolkit/training_extensions/pull/2227 was merged), and take a look on how it works for m-rcnn (both torch and OV tasks)? thanks!

File: CHANGELOG.md [hunk 7]
```diff
@@ -17,6 +17,7 @@ All notable changes to this project will be documented in this file.
 - Set persistent_workers and pin_memory as True in detection task (<https://github.com/openvinotoolkit/training_extensions/pull/2224>)
 - New algorithm for Semi-SL semantic segmentation based on metric lerning via class prototypes (https://github.com/openvinotoolkit/training_extensions/pull/2156)
 - Self-SL for classification now can recieve just folder with any images to start contrastive pretraining (https://github.com/openvinotoolkit/training_extensions/pull/2219)
+- Improve XAI saliency map generation for tiling detection (https://github.com/openvinotoolkit/training_extensions/pull/2240)
```
Comment (sovrasov) [inline CHANGELOG.md]: Could you please add https://github.com/openvinotoolkit/training_extensions/pull/2227 to changelog?
cc @negvet

**Comment by jaegukhyun** [comment]
Could you resolve conflicts?

File: otx/api/utils/tiler.py (hunk 5 above)
Comment (GalyaZalesskaya) [reply otx/api/utils/tiler.py]: Really good point, thank you

File: otx/algorithms/detection/adapters/mmdet/datasets/tiling.py (hunk 6 above)
Comment (GalyaZalesskaya) [reply otx/algorithms/detection/adapters/mmdet/datasets/tiling.py]: Fixed

**Comment by GalyaZalesskaya** [comment]
@eunwoosh @jaegukhyun @negvet @sovrasov 
Sorry for the delay, I was updating my PR to cover case with Mask RCNN Instance Segmentation model and to cover latest changes in develop.
Thank you all for your valuable comments. Can you please look again?

**Comment by GalyaZalesskaya** [comment]
@negvet @sovrasov Also would like to hear your opinion about moving saliency map generation for Mask RCNN to tiler.py

File: otx/api/utils/tiler.py [hunk 8]
```diff
@@ -270,6 +278,144 @@ def merge_results(self, results: List[Dict], shape: List[int]):
...
+            else:
+                image_map_cls = image_saliency_map[class_idx]
+                # resize the feature map for whole image to add it to merged saliency maps
+                if image_map_cls is not None:
+                    image_map_cls = cv2.resize(image_map_cls, (image_map_w, image_map_h))
+                    merged_map += (0.5 * image_map_cls).astype(dtype)
+                merged_map = non_linear_normalization(merged_map)
+        return merged_map
+
+    def get_tiling_saliency_map_from_segm_masks(self, detections: Union[Tuple, np.ndarray]) -> List:
```
Comment (sovrasov) [inline otx/api/utils/tiler.py]: I think, the openvino task should be also able to generate salienvy mask for MRCNN without tiling. Is that functionality preserved?

File: otx/api/utils/tiler.py [hunk 9]
```diff
@@ -270,6 +278,144 @@ def merge_results(self, results: List[Dict], shape: List[int]):
...
+        image_h, image_w, _ = image_meta["original_shape"]
+        ratio = np.array([feat_h, feat_w]) / self.tile_size
+
+        image_map_h = int(image_h * ratio[0])
+        image_map_w = int(image_w * ratio[1])
+        merged_map = [np.zeros((image_map_h, image_map_w)) for _ in range(num_classes)]
+
+        for (_, saliency_map), meta in features[1:]:
+            for class_idx in range(num_classes):
+                if saliency_map[class_idx] is None:
```
Comment (sovrasov) [inline otx/api/utils/tiler.py]: Can this happen? Is saliency map a list?

File: otx/api/utils/tiler.py (hunk 8 above)
Comment (GalyaZalesskaya) [reply otx/api/utils/tiler.py]: Yes, this functionality is saved and it's implemented in [openvino_models wrapper](https://github.com/openvinotoolkit/training_extensions/blob/42770dd6c6cbce5120fea08b46cefbd73d3ec5eb/otx/algorithms/detection/adapters/openvino/model_wrappers/openvino_models.py#L172) for Mask RCNN

File: otx/api/utils/tiler.py (hunk 9 above)
Comment (GalyaZalesskaya) [reply otx/api/utils/tiler.py]: Actually, you're right, it can't happened, I removed this altogether with the other checks if the saliency map for certain classes is None. 
The saliency map is `np.array` here, but I made the implementation more aligned with the implementation for PyTorch (tiling.py), where the saliency map can be a `np.array `(detection case) and a `list` (instance segmentation case).
To improve the performance, I removed extra iteration through classes and process saliency map as a `np.array`

**Review by sovrasov — DISMISSED** [review]

**Comment by eunwoosh** [comment]
Could you check fail test cases?

**Review by negvet — APPROVED** [review]

**Review by sovrasov — APPROVED** [review]
